week-2/README.md

```diff
@@ -0,0 +1,544 @@
+
+Run:
+
+```sh
+tofu apply --auto-approve
+```
+
+This will create the resources as defined in your configuration. Note that the database instance may take several minutes to become available.
+
+### 5. Verify the Deployment
+
+#### Step 5.1: Retrieve Outputs
+
+Run:
+
+```sh
+tofu output
+```
+
+You should see the public IP address of your EC2 instance and the endpoint of your database instance.
+
+Example output:
+
+```
+instance_public_ip = "54.123.45.67"
+db_endpoint        = "week2-db.abcdefghij.us-west-2.rds.amazonaws.com"
+```
+
+#### Step 5.2: Test the Instance
+
+Open a web browser and navigate to `http://<instance_public_ip>`. You should see the WordPress setup screen.
+
+### 6. Best Practices for Module Design
+
+- **Version Control**: Use versioning for your modules to manage changes and ensure stability.
+- **Documentation**: Provide clear documentation for inputs, outputs, and usage instructions.
+- **Minimal Exposure**: Only expose necessary variables to keep modules simple and user-friendly.
+- **Default Values**: Provide sensible defaults for variables where appropriate.
+- **Consistent Naming**: Use consistent naming conventions for resources and variables.
+- **Outputs**: Expose useful outputs to make the module more flexible.
+- **Error Handling**: Use validations to catch incorrect variable values.
+
+### 7. Cleaning Up
+
+Don't forget to destroy your infrastructure to avoid unnecessary charges:
+
+```sh
+tofu destroy -auto-approve
+```
+
+## Conclusion
+
+In this module, we expanded on the infrastructure from Week 1 by modularizing the `aws_instance` and `aws_db_instance` resources. We learned how to create reusable modules, manage dependencies between them, and apply best practices for module design.
+
+**Key Takeaways**:
+
+- **Modules**: Created custom modules for the `aws_instance` and `aws_db_instance`.
+- **Variables and Outputs**: Defined inputs and outputs to make modules flexible and reusable.
+- **Inter-Module Communication**: Used outputs from one module as inputs to another.
+- **Best Practices**: Applied best practices in module design for maintainability and collaboration.
+
+## Challenges
+
+Want to keep practicing before Week 3? Here are some challenges:
+
+1. **Enhance the Database Module**: Add [variables](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/db_instance) to configure more database options, such as backup retention, multi-AZ deployments, or storage type. Add [validations](https://opentofu.org/docs/language/expressions/custom-conditions/) to some of the variables for the database `variables.tf` (`instance_class`, `allocated_storage`, `engine`, `engine_version`)
+2. **Parameterize Security Groups**: Modify the security group definitions for `aws_instance` and `aws_db_instance` to accept lists of ports and protocols as variables.
+3. **Use AWS Secrets Manager**: Store the database password in AWS Secrets Manager and retrieve it in your configuration.
```

Comment:
Or SSM, Secrets Manager costs 40c a pop after 30 days, there were some notes from the previous challenge on slack.